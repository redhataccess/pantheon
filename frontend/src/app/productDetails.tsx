import React, { Component } from 'react';
import {
    ActionGroup, Breadcrumb, BreadcrumbItem, Button, Form, FormGroup, Level, LevelItem, List, ListItem,
    Text, TextContent, TextVariants, TextInput
} from '@patternfly/react-core';

export interface IProps {
    productName: string
}

class ProductDetails extends Component<IProps, any> {
    public versionNames: string[] = [];

    constructor(props) {
        super(props);
        this.state = {
            allVersionNames: [],
            newVersion: ''
        };
    }

    public render() {
        return (
            <React.Fragment>
                {this.state.fetchProductDetails && this.fetchProductDetails(this.state.allVersionNames)}
                <div className="app-container">
                    <Breadcrumb>
                        <BreadcrumbItem>All Products</BreadcrumbItem>
                        <BreadcrumbItem>Product Details</BreadcrumbItem>
                    </Breadcrumb>
                </div>
                <div className="app-container">
                    <Level gutter="md">
                        <LevelItem>
                            <TextContent>
                                <Text component={TextVariants.h1}>{this.props.productName}{'  '}</Text>
                            </TextContent>
                        </LevelItem>
                        <LevelItem />
                    </Level>
                </div>
                {this.state.allVersionNames.length !== 0 && (
                    <div className="app-container">
                        Versions:
                    <List>
                            {this.state.allVersionNames.map((version) =>
                                <ListItem key={version}>{version}</ListItem>
                            )}
                        </List>
                    </div>)}
                <div className="app-container" />
                <div className="app-container">
                    <Form>
                        <div className="app-container">
                            <FormGroup
                                label="New Version:"
                                fieldId="new_version_name"
                            >
                                <TextInput id="new_version_name_text" type="text" placeholder="New version name" onChange={this.handleTextInputChange} value={this.state.newVersion} />
                            </FormGroup>
                            <ActionGroup>
                                <Button aria-label="Creates a new Version Name." onClick={this.saveVersion}>Save</Button>
                            </ActionGroup>
                        </div>
                    </Form>
                </div>
            </React.Fragment>
        );
    }

    private fetchProductDetails = (versionNames) => {
        // setup url fragment
        const urlFragment = this.props.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_");
        const path = '/content/products/' + urlFragment + '/versions.2.json'
        let key;
        versionNames = []

        fetch(path)
            .then((response) => {
                if (response.ok) {
                    return response.json();
                } else if (response.status === 404) {
                    // create versions path
                    this.createVersionsPath();
                    return versionNames;
                } else {
                    throw new Error(response.statusText);
                }
            })
            .then(responseJSON => {
                // tslint:disable-next-line: prefer-for-of
                for (let i = 0; i < Object.keys(responseJSON).length; i++) {
                    key = Object.keys(responseJSON)[i];
                    if ((key !== 'jcr:primaryType')) {
                        if (responseJSON[key].name !== undefined) {
                            versionNames.push(responseJSON[key].name);
                        }
                    }
                }
                this.setState({
                    allVersionNames: versionNames
                })
            })
            .catch((error) => {
                console.log(error)
            });
        return versionNames;
    };

    private handleTextInputChange = newVersion => {
        this.setState({ newVersion });
    };

    private saveVersion = () => {
        const formData = new FormData();
        formData.append("name", this.state.newVersion)
        formData.append("sling:resourceType", "pantheon/productVersion")
        formData.append("jcr:primaryType", 'pant:productVersion')

        const urlFragment = this.props.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_");
        const encodedVersion = this.state.newVersion.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_");
        fetch(encodeURI('/content/products/' + urlFragment + '/versions/' + encodedVersion), {
            body: formData,
            method: 'post',
        }).then(response => {
            if (response.status === 200 || response.status === 201) {
                this.setState({ fetchProductDetails: true, newVersion: '' })
            } else {
                console.log('Version adding failure')
            }
        });

    };

    private createVersionsPath = () => {
        const formData = new FormData();
        const urlFragment = this.props.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_");
        fetch(encodeURI('/content/products/' + urlFragment + '/versions'), {
            body: formData,
            method: 'post',
        }).then(response => {
            if (response.status === 200 || response.status === 201) {
                console.log(" Created versions path " + response.status)
            } else {
                console.log(' Created versions path  failed!')

            }
        });
    }

}

export { ProductDetails }