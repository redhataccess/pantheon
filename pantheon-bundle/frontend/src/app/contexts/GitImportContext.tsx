import React, { createContext, useState, useEffect } from "react";

export let GitImportContext = createContext({} as any);

export interface IGitImportProviderProps {
  children?: React.ReactNode;
}

export interface IGitImport {
  id: number,
  repoName: string;
  totalFiles: number;
  success: boolean;
}

/**
 * Poll API until git import is completed
 */
const poll = async ({ fn, validate, interval }) => {
  let attempts = 0;
  const executePoll = async (resolve, reject) => {
    const result = await fn();
    attempts++;
    // we need to clone so we can read the response more than once with Response.json();
    const clonedResponse = result.clone();
    if (await validate(result)) {
      return resolve(clonedResponse);
    }
    //leaving this code commented out here in case we decide to add max attempts so API call doesn't last forever
    // else if (maxAttempts && attempts === maxAttempts) {
    //   return reject(new Error('Exceeded max attempts'));
    // } 
    else {
      setTimeout(executePoll, interval, resolve, reject);
    }
  };
  return new Promise(executePoll);
};

export function GitImportProvider({ children }: IGitImportProviderProps) {
  const [uploads, setUploads] = useState([] as IGitImport[]);

  const hdrs = {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
  };
  let git2pantheonURL
  let statusKey;
  let currRepository;

  // fn to clone git repo
  const cloneRepoGit2Pantheon = (branch, repository) => {
    currRepository = repository;
    const body = {
      'branch': branch,
      'repo': repository
    }
    fetch("/conf/pantheon/pant:syncServiceUrl")
    .then((resp => {
      if (!resp.ok) {
        console.log('Error occurred, could not find the git2pantheon URL configuration.');
        const singleUpload = {
          id: Date.now(),
          totalFiles: 0,
          repoName: repository,
          success: false
        };
        setUploads([...uploads, singleUpload]);
      } else {
        resp.text().then((text) => {
          git2pantheonURL= text
          console.log("The response text from pant:syncServiceUrl is: " + text)
        })
        .then(() => {
          if (git2pantheonURL === "") {
            console.log('Error occurred, the git2pantheon URL configuration is blank.')
            const singleUpload = {
              id: Date.now(),
              totalFiles: 0,
              repoName: repository,
              success: false
            };
            setUploads([...uploads, singleUpload]);
          } else {
    //first fetch call will return a status key as a response if successful 
    fetch(git2pantheonURL + "/api/clone", {
      body: JSON.stringify(body),
      headers: hdrs,
      method: 'POST'
    }).then(response => response.text().then(text => {
      statusKey = text;
      //if first fetch fails, upload has failed a return a single upload with 0 total files
      if (text.includes("error")) {
        console.log('ERROR! in /clone');
        const singleUpload = {
          id: Date.now(),
          totalFiles: 0,
          repoName: repository,
          success: false
        };
        setUploads([...uploads, singleUpload]);
      } else {
        const validateStatus = async (response) => {
          const json = await response.json();
          return json.status === "done" || json.status === "error";
        }
        //second fetch takes the status key and returns the status of repo upload once it is complete
        //we keep making this call until status === "done" signifying the upload has completed
        const getStatus = async () => {
          return await fetch(git2pantheonURL + "/api/status", {

            body: statusKey,
            headers: hdrs,
            method: 'POST'
          })
        };
        poll({
          fn: getStatus,
          validate: validateStatus,
          interval: 2000
        }).then((response: any) => {
          response.json().then(json => {
            let singleUpload
            if (json.status === "done") {
              singleUpload = {
                id: Date.now(),
                totalFiles: json.total_files_uploaded,
                repoName: repository,
                success: true
              };
            }
            else if (json.status === "error") {
              singleUpload = {
                id: Date.now(),
                totalFiles: 0,
                repoName: repository,
                success: false
              };
            }
            setUploads([...uploads, singleUpload]);
          });
        });
      }
    })).catch(e => console.error("ERROR! in /status", e));
  }
        })
        }
      }))
  }

  // components wrapped in this provider will now have access to uploads and cloneRepoGit2Pantheon
  return <GitImportContext.Provider value={{ uploads, cloneRepoGit2Pantheon }}>
    {children}
  </GitImportContext.Provider>
}
