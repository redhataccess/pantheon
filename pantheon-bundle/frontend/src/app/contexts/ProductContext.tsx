import React, { createContext, useState, useEffect } from "react";

export let ProductContext = createContext<IProduct[]>([] as IProduct[]);

export interface IProductProviderProps {
  children?: React.ReactNode;
}

export interface IProduct {
  description: string,
  isOpen: boolean,
  ["jcr:created"]: Date,
  ["jcr:createdBy"]: string,
  ['jcr:lastModified']: Date,
  ["jcr:lastModifiedBy"]: string,
  ["jcr:primaryType"]: string,
  ["jcr:uuid"]: string,
  locale: string
  name: string,
  ['sling:resourceType']: string,
  urlFragment: string,
}

export function ProductProvider({ children }: IProductProviderProps) {
  const backend = "/content/products.query.json?nodeType=pant:product&orderby=name"
  const [allProducts, setAllProducts]: any = useState([] as IProduct[]);

  useEffect(() => {
    try {
      // fetch all product data and set allProducts variable to the response
      fetch(backend)
        .then(response => response.json())
        .then(response => {
          response.results.forEach(product => {
            product.isOpen = false
          })
          setAllProducts(response.results);
        }).catch(e => console.error(e));
    } catch {
      console.error("unexpected error");
    }

  }, []);
  // components wrapped in this provider will now have access to allProducts
  return <ProductContext.Provider value={allProducts}>
    {children}
  </ProductContext.Provider>
}
