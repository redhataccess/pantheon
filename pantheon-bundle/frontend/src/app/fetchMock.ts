if (!self.fetch) {
    self.fetch = (input: RequestInfo, init?: RequestInit | undefined) => {
        const p = new Promise<Response>((resolve, reject) => {
            // just don't do anything. this exists so that our tests don't try and actually resolve fetches,
            // (which is exactly what isomorphic-fetch tried to do while we were importing it...)
            // and calling nothing here allows us to skip all of our fetch.then() and fetch.catch() code.
        })
        return p
    }
}