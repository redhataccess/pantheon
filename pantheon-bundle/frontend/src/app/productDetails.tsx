import React, { useState, useEffect, useContext } from "react"
import {
    ActionGroup, Alert, AlertActionCloseButton, Breadcrumb, BreadcrumbItem, Button, Form, FormGroup, Level, LevelItem, List, ListItem,
    Text, TextContent, TextVariants, TextInput, InputGroup
} from "@patternfly/react-core"
import { Fields, JcrTypes, SlingTypes } from "@app/Constants"
import { ProductContext, IProduct } from "@app/contexts/ProductContext"


export const ProductDetails = (props: any) => {
    const allProducts = useContext(ProductContext);

    const [failedPost, setFailedPost] = useState(false)
    const [isMissingFields, setIsMissingFields] = useState(false)
    const [urlFragment, setUrlFragment] = useState("")
    const [isUrlFragmentValid, setIsUrlFragmentValid] = useState(true)
    const [newVersion, setNewVersion] = useState("")
    const [allVersionNames, setAllVersionNames] = useState([] as string[])
    const [currentProduct, setCurrentProduct] = useState(undefined as IProduct | undefined)

    useEffect(() => {
        if (allProducts) {
            const currentProduct = allProducts.find(a => a["jcr:uuid"] === props.match.params.id)
            setCurrentProduct(currentProduct)
            fetchProductVersions(currentProduct)
        }
    }, [allProducts])

    const fetchProductVersions = (currentProduct) => {
        const urlFragment = currentProduct && currentProduct.name.toLowerCase().replace(/[^A-Z0-9]+/ig, "_")
        const path = "/content/products/" + urlFragment + "/versions.2.json"
        let versionNames: string[] = []
        fetch(path).then((response) => {
            if (response.ok) {
                return response.json()
            } else if (response.status === 404) {
                // create versions path
                createVersionsPath(urlFragment)
                return versionNames
            } else {
                throw new Error(response.statusText)
            }
        }).then(json => {
            for (let i = 0; i < Object.keys(json).length; i++) {
                const key = Object.keys(json)[i]
                if (key !== Fields.JCR_PRIMARYTYPE && json[key].name !== undefined) {
                    versionNames.push(json[key].name)
                }
            }
            setAllVersionNames(versionNames)
        })
            .catch((error) => {
                console.log(error)
            })
    }

    const createVersionsPath = (urlFragment) => {
        const formData = new FormData()
        fetch(encodeURI("/content/products/" + urlFragment + "/versions"), {
            body: formData,
            method: "post",
        }).then(response => {
            if (response.status === 200 || response.status === 201) {
                console.log(" Created versions path " + response.status)
            } else {
                console.log(" Created versions path failed!")
            }
        })
    }

    const dismissNotification = () => {
        setFailedPost(false)
        setIsMissingFields(false)
    }

    const handleTextInputChange = newVersion => {
        setNewVersion(newVersion)
    }

    const handleUrlInputChange = urlFragment => {
        if (/^[-.\w]+$/.test(urlFragment)) {
            setUrlFragment(urlFragment)
            setIsUrlFragmentValid(true)
        } else {
            setIsUrlFragmentValid(false)
        }
    }

    const saveVersion = () => {
        if (newVersion === "" || urlFragment === "") {
            setIsMissingFields(true)
        } else {
            // creating formData obj and adding form's data to it
            const formData = new FormData()
            formData.append(Fields.NAME, newVersion)
            formData.append(Fields.SLING_RESOURCETYPE, SlingTypes.PRODUCT_VERSION)
            formData.append(Fields.JCR_PRIMARYTYPE, JcrTypes.PRODUCT_VERSION)
            formData.append(Fields.URL_FRAGMENT, urlFragment)

            // creating url-friendly version of product name and product version
            const productUrlFragment = currentProduct && currentProduct.name.toLowerCase().replace(/[^A-Z0-9]+/ig, "_")
            const encodedVersion = newVersion.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_")
            // sending formData to backend
            fetch(encodeURI("/content/products/" + productUrlFragment + "/versions/" + encodedVersion), {
                body: formData,
                method: "post",
            }).then(response => {
                if (response.status === 200 || response.status === 201) {
                    setNewVersion("")
                    setUrlFragment("")
                    currentProduct && fetchProductVersions(currentProduct)
                } else {
                    setFailedPost(true)
                    setIsMissingFields(false)
                    console.log("Version adding failure")
                }
            })
        }
    }

    return (
        <React.Fragment>
            <div className="app-container">
                <Breadcrumb>
                    <BreadcrumbItem ><a href="#/products">All Products</a></BreadcrumbItem>
                    <BreadcrumbItem to="#" isActive={true}>Product Details</BreadcrumbItem>
                </Breadcrumb>
            </div>
            {currentProduct && <div className="app-container">
                <Level>
                    <LevelItem>
                        <TextContent>
                            <Text component={TextVariants.h1}>{currentProduct.name}{"  "}</Text>
                        </TextContent>
                    </LevelItem>
                    <LevelItem />
                </Level>
            </div>}
            {allVersionNames.length !== 0 && (
                <div className="app-container">
                    Versions:
                    <List>
                        {allVersionNames.map((version) =>
                            <ListItem key={version}>{version}</ListItem>
                        )}
                    </List>
                </div>)}
            <br />
            <div>
                <Form>
                    <div className="pant-form-width-md">
                        {isMissingFields &&
                            <div className="notification-container">
                                <Alert variant="warning"
                                    title="Fields indicated by * are mandatory"
                                    actionClose={<AlertActionCloseButton onClose={dismissNotification} />}
                                />
                            </div>
                        }
                        {!isUrlFragmentValid &&
                            <div className="notification-container">
                                <Alert variant="warning"
                                    title="Allowed input for ulrFragment: alphanumeric, hyphen, period and underscore"
                                    actionClose={<AlertActionCloseButton onClose={dismissNotification} />}
                                />
                            </div>
                        }
                        <FormGroup
                            label="New Version:"
                            isRequired={true}
                            fieldId="new_version_name"
                        >
                            <TextInput id="new_version_name_text" type="text" placeholder="New version name" onChange={handleTextInputChange} value={newVersion} />
                        </FormGroup>
                        <br />
                        <FormGroup
                            label=" URL Fragment:"
                            isRequired={true}
                            fieldId="url_fragment"
                        >
                            <TextInput id="url_fragment_text" type="text" placeholder="Url fragment" onChange={handleUrlInputChange} value={urlFragment} />
                        </FormGroup>
                        {failedPost &&
                            <div className="notification-container">
                                <Alert
                                    variant="danger"
                                    title="Failed to create product version."
                                    actionClose={<AlertActionCloseButton onClose={dismissNotification} />}
                                >
                                    Please check if you are logged in as a publisher.
                                    </Alert>
                            </div>
                        }
                        {allProducts && allProducts.length > 0 &&
                            <ActionGroup>
                                <Button aria-label="Creates a new Version Name." onClick={saveVersion}>Save</Button>
                            </ActionGroup>
                        }
                    </div>
                </Form>
            </div>
        </React.Fragment>
    )
}
