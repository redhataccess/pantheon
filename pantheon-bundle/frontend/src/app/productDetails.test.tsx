import React from "react"
import "@app/fetchMock"
import { mount, shallow } from "enzyme"
import { Breadcrumb, Button, Form, FormGroup, Level, LevelItem, TextContent, TextInput, Text } from "@patternfly/react-core"
import { ProductDetails } from "@app/productDetails"
import { ProductContext, IProduct } from "./contexts/ProductContext"


const date = new Date();
const allProducts: IProduct[] = [{
  description: "descrip",
  isOpen: false,
  ["jcr:created"]: date,
  ['jcr:createdBy']: "admin",
  ['jcr:lastModified']: date,
  ['jcr:lastModifiedBy']: "admin",
  ['jcr:primaryType']: "pant:product",
  ['jcr:uuid']: "5df8d913-79b9-42fd-a17b-ffc917add446",
  locale: "en-US",
  name: "product 11/16",
  ['sling:resourceType']: "pantheon/product",
  urlFragment: "product_url",
}, {
  description: "test descrip",
  isOpen: false,
  ['jcr:created']: date,
  ['jcr:createdBy']: "admin",
  ['jcr:lastModified']: date,
  ['jcr:lastModifiedBy']: "admin",
  ['jcr:primaryType']: "pant:product",
  ['jcr:uuid']: "242d4187-d2f4-4df5-917b-e09bf4ff45e9",
  locale: "en-US",
  name: "test 11/17",
  ['sling:resourceType']: "pantheon/product",
  urlFragment: "testurl",
}];

const props =
{
  match:
  {
    params: {
      id: '5df8d913-79b9-42fd-a17b-ffc917add446'
    }
  }
}
const container = mount(<ProductContext.Provider value={allProducts}><ProductDetails {...props} /></ProductContext.Provider>)

describe("ProductDetails tests", () => {
  test("should render ProductDetails component", () => {
    const view = shallow(<ProductDetails {...props} />)
    expect(view).toMatchSnapshot()
  })

  it("should render a form", () => {
    const form = container.find(Form)
    expect(form.exists()).toBe(true)
  })

  it("should render a form group", () => {
    const formGroup = container.find(FormGroup)
    expect(formGroup.exists()).toBe(true)
  })

  it("should render a text input", () => {
    const textInput = container.find(TextInput)
    expect(textInput.exists()).toBe(true)
  })

  it("should render a Button", () => {
    const button = container.find(Button)
    expect(button.exists()).toBe(true)
  })

  it("should render a breadcrumb", () => {
    const breadCrumb = container.find(Breadcrumb)
    expect(breadCrumb.exists()).toBe(true)
  })

  it("should render a Level element", () => {
    const level = container.find(Level)
    expect(level.exists()).toBe(true)
  })

  it("should render a LevelItem", () => {
    const levelItem = container.find(LevelItem)
    expect(levelItem.exists()).toBe(true)
  })

  it("should render a TextContent element", () => {
    const textContent = container.find(TextContent)
    expect(textContent.exists()).toBe(true)
  })

  it("should render a Text element", () => {
    const text = container.find(Text)
    expect(text.exists()).toBe(true)
  })
})

describe('fetchProductVersions', () => {
  it('fetches data and returns a successful response', done => {
    const mockSuccessResponse = {};
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    mount(<ProductContext.Provider value={allProducts}><ProductDetails {...props} /></ProductContext.Provider>)
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith('/content/products/product_11_16/versions.2.json');
    done();
  });
});
