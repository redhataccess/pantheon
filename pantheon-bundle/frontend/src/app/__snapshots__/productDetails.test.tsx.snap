// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails tests should render ProductDetails component 1`] = `
<Fragment>
  <div
    className="app-container"
  >
    <Breadcrumb>
      <BreadcrumbItem>
        <a
          href="#/products"
          onClick={[Function]}
        >
          All Products
        </a>
      </BreadcrumbItem>
      <BreadcrumbItem
        isActive={true}
        to="#"
      >
        Product Details
      </BreadcrumbItem>
    </Breadcrumb>
  </div>
  <div
    className="app-container"
  >
    <Level>
      <LevelItem>
        <TextContent>
          <Text
            component="h1"
          >
            Red Hat Enterprise Linux
              
          </Text>
        </TextContent>
      </LevelItem>
      <LevelItem />
    </Level>
  </div>
  <br />
  <div>
    <Form>
      <div
        className="pant-form-width-md"
      >
        <FormGroup
          fieldId="new_version_name"
          isRequired={true}
          label="New Version:"
        >
          <ForwardRef
            id="new_version_name_text"
            onChange={[Function]}
            placeholder="New version name"
            type="text"
            value=""
          />
        </FormGroup>
        <br />
        <FormGroup
          fieldId="url_fragment"
          isRequired={true}
          label=" URL Fragment:"
        >
          <ForwardRef
            id="url_fragment_text"
            onChange={[Function]}
            placeholder="Url fragment"
            type="text"
            value=""
          />
        </FormGroup>
        <ActionGroup>
          <Button
            aria-label="Creates a new Version Name."
            onClick={[Function]}
          >
            Save
          </Button>
        </ActionGroup>
      </div>
    </Form>
  </div>
</Fragment>
`;

exports[`ProductDetails tests test componentWillReceiveProps function 1`] = `undefined`;

exports[`ProductDetails tests test createVersionsPath function 1`] = `[Function]`;

exports[`ProductDetails tests test fetchProductDetails function 1`] = `Array []`;

exports[`ProductDetails tests test handleUrlInputChange function 1`] = `undefined`;

exports[`ProductDetails tests test saveVersion function 1`] = `[Function]`;
